### PR TITLE
5748 Filter out nested serializer fields in the API and defer the filtered fields.

### PR DESCRIPTION
**Pull request by albertisfu**
This PR leverages the new mixin `NestedDynamicFieldsMixin`  for `drf-dynamic-fields` to enable filtering of nested fields, as introduced in https://github.com/dbrgn/drf-dynamic-fields/pull/44

This allows requests like:
`GET /api/rest/v4/docket-entries/?fields=id,recap_documents__plain_text,recap_documents__id
`
to filter both parent and nested fields:

Screenshot 2025-07-14 at 7 54 12 p m

Similarly, the `omit` parameter can be used:
`GET /api/rest/v4/docket-entries/?omit=entry_number,recap_documents__plain_text`

Screenshot 2025-07-14 at 7 55 02 p m

This mixin has been added to parent serializers that render nested objects and to their respective child serializers:

**Docket Entries:**

- `DocketEntrySerializer`
- `RECAPDocumentSerializer`

**Financial Disclosures:**

- `FinancialDisclosureSerializer`
- `AgreementSerializer`
- `DebtSerializer`
- `InvestmentSerializer`
- `GiftSerializer`
- `NonInvestmentIncomeSerializer`
- `PositionSerializer`
- `ReimbursementSerializer`
- `SpouseIncomeSerializer`

**People:**

- `PersonSerializer`
- `EducationSerializer`

### Deferred fields.
To optimize query performance by deferring filtered fields from the request, this PR introduces two additional mixins:

**`RetrieveFilteredFieldsMixin`**
Retrieves the filtered fields from the `fields` or `omit` query parameters, allowing the `ViewSet` to access them and defer them accordingly.

**`DeferredFieldsMixin`**
Overrides the `get_queryset` method to defer fields based on the filtered request fields. Key behaviors:

- **Conflict with `select_related`:**
    You cannot defer a field that is also included in `select_related`. When this happens, the `select_related` field is removed from the deferred list.
    Prioritizing the deferred field by removing the `select_related` entry is not ideal, because serializers may reference that relation indirectly (e.g., via a property), making it hard to predict whether it's needed. For example in the `docket-entries` endpoint, `select_related("docket")` is used. The `docket` field is referenced directly in the serializer and also indirectly in `RECAPDocument.get_absolute_url`.
    
- **Original deferred fields preserved:**
    Any fields already deferred in the original `queryset` are preserved and combined with user-defined ones.
    
- **`only()`:*
    Django prioritizes `only()` over `defer()`. If the original queryset uses `only()`, user-specified fields to defer are ignored.
    
- **Ordering `prefetch_related`:**
    Custom `Prefetch` objects (used for nested field deferring) must appear before the original `prefetch_related()` lookups to avoid conflicts.
    
- **Nested relationship keys:**
    Foreign key fields that link nested objects to their parent must not be deferred, even if omitted in the response. Deferring them would trigger extra queries.
    For example, `recap_documents__docket_entry_id` is required for joining RECAPDocuments to DocketEntries, even if omitted from the payload. If the user includes this field in the omit list, it is removed from the payload but ignored during the deferring process.
    
- **No support for `.values()` or `.annotate()`:**
    If the base queryset uses `.values()` or annotations, the deferring logic is ignored, as our current ViewSets were this logic was applied don’t use them.

Here an example of how `DeferredFieldsMixin` works for the request:
`api/rest/v4/docket-entries/?fields=id,recap_documents__plain_text,recap_documents__id`

`DocketEntry` (parent) selects only the id field. However, due to select_related("docket"), all docket fields are also selected:

Screenshot 2025-07-14 at 8 32 50 p m

`RECAPDocuments` (nested) select only the `id`, `plain_text`, and `docket_entry_id` fields:
Screenshot 2025-07-14 at 8 34 01 p m

In this PR I've added these mixins `RetrieveFilteredFieldsMixin` and `DeferredFieldsMixin` to the following viewsets:

AudioViewSet

DocketViewSet
DocketEntryViewSet
RECAPDocumentViewSet
CourtViewSet
OpinionClusterViewSet
OpinionViewSet
OpinionsCitedViewSet
TagViewSet
OriginatingCourtInformationViewSet

FinancialDisclosureViewSet
AgreementViewSet
DebtViewSet
GiftViewSet
InvestmentViewSet
NonInvestmentIncomeViewSet
PositionViewSet
ReimbursementViewSet
SpouseIncomeViewSet

PersonViewSet
PositionViewSet
RetentionEventViewSet
EducationViewSet
SchoolViewSet
PoliticalAffiliationViewSet
SourceViewSet
ABARatingViewSet

For now, `drf-dynamic-fields` is temporarily installed from my forked repository. Once the upstream PR is merged and a new release is published, we can update the dependency to point back to the official package.

Let me know what do you think.

### CODE REVIEW

File: cl/api/utils.py [hunk 1]
```diff
@@ -1218,3 +1218,245 @@ def handle_webhook_events(results: list[int | float], user: User) -> None:
...
+        except KeyError:
+            conf = getattr(settings, "DRF_DYNAMIC_FIELDS", {})
+            if conf.get("SUPPRESS_CONTEXT_WARNING", False) is not True:
+                logger.warning(
+                    "Context does not have access to request. "
+                    "See README for more information."
+                )
+
+        # NOTE: drf test framework builds a request object where the query
+        # parameters are found under the GET attribute.
```
Comment (ERosendo) [inline cl/api/utils.py]: From what I understand, it's the default behavior of Django’s [HttpRequest](https://github.com/django/django/blob/6e71386118b1e9af0bf60b75ad9549aa5df2192d/django/http/request.py#L53) class to store query parameters as a dictionary in the GET attribute. It’s actually DRF that adds a [wrapper](https://github.com/encode/django-rest-framework/blob/853969c69c815be69513c2f63a41285858a45352/rest_framework/request.py#L140) around Django’s request to enhance it.

That said, under the hood, query_params just returns self._request.GET ([source](https://github.com/encode/django-rest-framework/blob/853969c69c815be69513c2f63a41285858a45352/rest_framework/request.py#L211)). So I think it would be fine to just use request.GET directly in this context.

File: cl/api/utils.py [hunk 2]
```diff
@@ -1218,3 +1218,245 @@ def handle_webhook_events(results: list[int | float], user: User) -> None:
...
+            if "__" in filtered_field:
+                parent, child = filtered_field.split("__", 1)
+                self._nested_allow[parent].append(child)
+                # If a nested field is allowed the related parent level field
+                # must also be allowed
+                self._flat_allow.add(parent)
+            else:
+                self._flat_allow.add(filtered_field)
+
+            # store top-level and nested fields in the `omit` argument.
```
Comment (ERosendo) [inline cl/api/utils.py]: Looks like the indentation of this comment could use a quick fix.

File: cl/api/utils.py [hunk 3]
```diff
@@ -1218,3 +1218,245 @@ def handle_webhook_events(results: list[int | float], user: User) -> None:
...
+            if nested_model is None:
+                continue
+
+            # Filter out nested fields that have a database column associated
+            # with them.
+            field_names = [
+                f.name
+                for f in nested_model._meta.get_fields()
+                if getattr(f, "concrete", False)
+            ]
```
Comment (ERosendo) [inline cl/api/utils.py]: This list comprehension could be refactored to avoid repetition; it’s similar to the one in `_filter_top_level_fields_to_defer`.

File: cl/api/utils.py [hunk 4]
```diff
@@ -1218,3 +1218,245 @@ def handle_webhook_events(results: list[int | float], user: User) -> None:
...
+        Determine which top-level model fields should be deferred when an
+        explicit omit filter is in use. Valid database fields in the omit list
+        will be deferred.
+        """
+        omit = getattr(self, "_flat_omit", None)
+        return self._filter_top_level_fields_to_defer(
+            omit, keep_if=lambda name, omit: name in omit
+        )
+
+    def _get_nested_level_fields_to_defer(self, nested_mapping, should_defer):
```
Comment (ERosendo) [inline cl/api/utils.py]: Correct me if I’m wrong, but it looks like this method only supports a single level of nesting, right? I also checked the tests and noticed that the maximum depth of the fields we’re deferring is one level as well.

File: cl/api/utils.py [hunk 5]
```diff
@@ -1218,3 +1218,245 @@ def handle_webhook_events(results: list[int | float], user: User) -> None:
...
+        for parent_field, child_fields in nested_map.items():
+            field = serializer.fields.get(parent_field)
+            if not field:
+                continue
+
+            child_serializer = getattr(field, "child", field)
+            child_model = getattr(child_serializer.Meta, "model", None)
+            parent_model = serializer.Meta.model
+            if not child_model or not parent_model:
+                continue
```
Comment (ERosendo) [inline cl/api/utils.py]: This section feels a bit redundant, as we already run similar logic in the `_get_nested_level_fields_to_defer` method of the `RetrieveFilteredFieldsMixin` class.

Also, since this mixin is always used alongside `RetrieveFilteredFieldsMixin`, the `nested_map` variable should already contain a cleaned version of the fields derived from the `field` and `omit` query parameters.

File: cl/api/utils.py (hunk 4 above)
Comment (albertisfu) [reply cl/api/utils.py]: Yeah, field deferring is supported for top‑level fields and one nested level, which covers our serializer use cases, for example, the `docket-entries` endpoint that nests RECAPDocument objects. A few serializers go to a third level of nesting, but they use `HyperlinkedRelatedField` which is just a link. Only `PersonSerializer` is the only one that has 2 levels of nested objects. So I think that supporting multiple levels of deferred fields would be significantly more complex, so this implementation already satisfies the majority of our serializers, and cover the initial request per: https://github.com/freelawproject/courtlistener/issues/5748

I think that if one day we have the a performance issue related to a serializer with more than 1 nested level, we can then extend the deferring logic. 

Let me know what do you think.

Comment (ERosendo) [reply cl/api/utils.py]: > I think that if one day we have the a performance issue related to a serializer with more than 1 nested level, we can then extend the deferring logic.

Thanks for the explanation. That makes sense 👍

Comment (mlissner) [reply cl/api/utils.py]: I've always hesitated to go to that third level. It starts to feel like you're asking too much of a single API response, but maybe I'm silly.

File: cl/api/utils.py (hunk 1 above)
Comment (albertisfu) [reply cl/api/utils.py]: Got it! You're right. I've simplified this code.

File: cl/api/utils.py (hunk 3 above)
Comment (albertisfu) [reply cl/api/utils.py]: Sure I've introduced the new method `_get_concrete_fields_for_model` for this.

File: cl/api/utils.py (hunk 5 above)
Comment (albertisfu) [reply cl/api/utils.py]: Yeah, I've removed the unnecessary checks already performed in `RetrieveFilteredFieldsMixin`

**Comment by albertisfu** [comment]
Thanks @ERosendo for your review. I've applied your suggestions. 
I also added type hints to the methods and resolved the mypy conflicts, this is ready for another review.

File: cl/api/utils.py (hunk 5 above)
Comment (albertisfu) [reply cl/api/utils.py]: I’ve updated the approach in `_get_nested_level_fields_to_defer` so that it now returns a dictionary of field objects. These are then merged in `get_deferred_model_fields` and used directly in `DeferredFieldsMixin`, eliminating redundant steps.

Let me know what you think. Thank you!

**Comment by albertisfu** [comment]
Once this PR is ready, we’ll need to decide what to do with https://github.com/dbrgn/drf-dynamic-fields/pull/44 before merging this one. Since we haven’t received a response from the maintainer. Do you a suggestion @mlissner ?

**Comment by mlissner** [comment]
I think we've done all we can there. Our choice is to fork it or leave it? I'm not sure we have many other options. Seems like we've solved our immediate problems with the approach you have here?

**Comment by albertisfu** [comment]
> I think we've done all we can there. Our choice is to fork it or leave it? I'm not sure we have many other options. Seems like we've solved our immediate problems with the approach you have here?

That’s correct. However, this PR depends on the modified `DynamicFieldsMixin` from https://github.com/dbrgn/drf-dynamic-fields/pull/44 to filter nested fields in the response. Right now it pulls from my forked repo https://github.com/albertisfu/drf-dynamic-fields, which might not be ideal for production. We could instead fork the upstream repo or integrate the entire `DynamicFieldsMixin` directly into the CL code?

**Comment by mlissner** [comment]
Oh shoot, I forgot that. I don't really want to fork it and deal with python packaging. If it's just the mixin, let's put it into CL directly?

**Comment by albertisfu** [comment]
Sure. In total, it will include the modified `DynamicFieldsMixin` and a new `NestedDynamicFieldsMixin`, comprising about 200 lines of code

**Comment by albertisfu** [comment]
@ERosendo  I've moved `DynamicFieldsMixin` and the `NestedDynamicFieldsMixin` to cl.api.utils and removed `drf_dynamic_fields` dependency.

**Comment by mlissner** [comment]
Fantastic work, guys. Since we're removing the dependency, should we make a note in the mixin (or somewhere else) that we can remove the mixin when the PR is merged (if it ever is)? I glanced but didn't see this.
